spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,181 @@
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{sources, DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnectorConf, CassandraConnector, Schema, ColumnDef}
+import com.datastax.spark.connector.rdd.{CassandraRDD, ReadConf}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+import DataTypeConverter._
+
+/**
+ * Implements [[BaseRelation]]]], [[InsertableRelation]]]] and [[PrunedFilteredScan]]]]
+ * It inserts data to and scans Cassandra table. If filterPushdown is true, it pushs down
+ * some filters to CQL
+ *
+ */
+private[cassandra] class CassandraSourceRelation(
+    tableRef: TableRef,
+    userSpecifiedSchema: Option[StructType],
+    filterPushdown: Boolean,
+    tableSizeInBytes: Long,
+    connector: CassandraConnector,
+    readConf: ReadConf,
+    writeConf: WriteConf,
+    override val sqlContext: SQLContext)
+  extends BaseRelation
+  with InsertableRelation
+  with PrunedFilteredScan
+  with Logging {
+
+  private[this] val tableDef = Schema.fromCassandra(connector)
+    .keyspaceByName(tableRef.keyspace).tableByName(tableRef.table)
+
+  override def schema: StructType = {
+    userSpecifiedSchema.getOrElse(StructType(tableDef.allColumns.map(toStructField)))
+  }
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    if (overwrite) {
+      connector.withSessionDo {
+        val keyspace = quoted(tableRef.keyspace)
+        val table = quoted(tableRef.table)
+        session => session.execute(s"TRUNCATE $keyspace.$table")
+      }
+    }
+
+    implicit val rwf = SqlRowWriter.Factory
+    data.rdd.saveToCassandra(tableRef.keyspace, tableRef.table, AllColumns, writeConf)
+  }
+
+  override def sizeInBytes: Long = tableSizeInBytes
+
+  implicit val cassandraConnector = connector
+  implicit val readconf = readConf
+  private[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](tableRef.keyspace, tableRef.table)
+
+  def buildScan() : RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    logInfo(s"filters: ${filters.mkString(", ")}")
+    val pushdownFilters = if(filterPushdown) new PushDown(filters, tableDef).toPushDown else Seq.empty
+    if(filterPushdown) {
+      logInfo(s"pushdown filters: ${pushdownFilters.toString()}")
+    }
+
+    val transformer = new RddTransformer(requiredColumns, pushdownFilters)
+    transformer.transform(baseRdd).asInstanceOf[RDD[Row]]
+  }
+
+  /** Quote name */
+  private[this] def quoted(str: String): String = {
+    "\"" + str + "\""
+  }
+
+  /** Add column selection and where clauses on top of baseRdd */
+  private[this] class RddTransformer(requiredColumns: Array[String], filters: Seq[Filter]) {
+
+    /** CassandraRDD[CassandraSQLRow] is the only type supported for transferring */
```

Comment:
Limit the type only to CassandraRDD[CassandraSQLRow]. Change it to "Define a type for CassandraRDD[CassandraSQLRow]. It's used by following methods."